src/Haskell/Ide/Engine/Plugin/HsImport.hs

```diff
@@ -79,25 +75,51 @@ importModule uri modName =
         Nothing -> do
           newText <- liftIO $ T.readFile output
           liftIO $ removeFile output
-          J.WorkspaceEdit mChanges mDocChanges <- liftToGhc $ makeDiffResult input newText fileMap
+          J.WorkspaceEdit mChanges mDocChanges <- liftToGhc
+            $ makeDiffResult input newText fileMap
 
           if shouldFormat
             then do
-              -- Format the import with Brittany
-              confFile <- liftIO $ Brittany.getConfFile origInput
-              newChanges <- forM mChanges $ mapM $ mapM (formatTextEdit confFile)
-              newDocChanges <- forM mDocChanges $ mapM $ \(J.TextDocumentEdit vDocId tes) -> do
-                ftes <- forM tes (formatTextEdit confFile)
-                return (J.TextDocumentEdit vDocId ftes)
-
-              return $ IdeResultOk (J.WorkspaceEdit newChanges newDocChanges)
-            else
-              return $ IdeResultOk (J.WorkspaceEdit mChanges mDocChanges)
-
-  where formatTextEdit confFile (J.TextEdit r t) = do
-          -- TODO: This tab size of 2 spaces should probably be taken from a config
-          ft <- fromRight t <$> liftIO (Brittany.runBrittany 2 confFile t)
-          return (J.TextEdit r ft)
+              config  <- getConfig
+              plugins <- getPlugins
+              let mprovider = Hie.getFormattingPlugin config plugins
+              case mprovider of
+                Nothing ->
+                  return $ IdeResultOk (J.WorkspaceEdit mChanges mDocChanges)
+
+                Just (plugin, _) -> do
+                  newChanges <- forM mChanges $ \change -> do
+                    let func = mapM (formatTextEdit plugin)
+                    res <- mapM func change
+                    return $ fmap flatten res
+
+                  newDocChanges <- forM mDocChanges $ \change -> do
+                    let cmd (J.TextDocumentEdit vids edits) = do
+                          newEdits <- mapM (formatTextEdit plugin) edits
+                          return $ J.TextDocumentEdit vids (flatten newEdits)
+                    mapM cmd change
+
+                  return
+                    $ IdeResultOk (J.WorkspaceEdit newChanges newDocChanges)
+            else return $ IdeResultOk (J.WorkspaceEdit mChanges mDocChanges)
+
+ where
+  flatten :: List [a] -> List a
+  flatten (J.List list) = J.List (join list)
+
+  formatTextEdit :: PluginDescriptor -> J.TextEdit -> IdeGhcM [J.TextEdit]
+  formatTextEdit plugin edit@(J.TextEdit r t) = do
+    result <- runPluginCommand
+      (pluginId plugin)
+      "formatText"
+      -- TODO: should this be in the configs?
+      (dynToJSON $ toDynJSON $ FormatTextCmdParams t
+                                                   r
+                                                   (FormattingOptions 2 True)
```

Comment:
How could we encode this in a way that it is available to a IdeGhcM?

Reply:
I think we can just use the `formatting provider` directly here no? Since Hie.getFormattingPlugin should return both the plugin and the `FormattingProvider`, which is just a function.

Reply:
FomattingProvider works in IdeDeferM. Can this be integrated into IdeGhcM? I could not find a way.

Reply:
Moreover, R doesnt know the `FormattingOptions` either.